File: src/js/main/RichJsonHelper.js
import {__RICH_JSON_CLASS_MAPPING} from "./RichJsonClassMapping.js";
import {RichJsonCache} from "./RichJsonCache.js";
import {__RICH_JSON_CONFIG} from "./RichJsonConfiguration.js";

/**
 * Merges two or more objects together into a new object.
 * @param {object} objects Going to be merged
 * @returns {object} The merged object
 */
export function mergeObjects(...objects) {
    let rv = {};
    let cache = new RichJsonCache();
    for (let i = 0; i < objects.length; i++) {
        _mergeIntoTarget(cache, rv, [objects[i]]);
    }
    if (__RICH_JSON_CONFIG.debugEnabled && cache.level !== 0) {
        console.error(`RichJson mergeIntoTarget failed!`);
    }
    return rv;
}

/**
 * Merges objects into target object.
 * @param {object} target
 * @param {object} others Going to be merged
 * @returns {object} The merged target
 */
export function mergeIntoTarget(target, ...others) {
    let cache = new RichJsonCache();
    target = _mergeIntoTarget(cache, target, others);
    if (__RICH_JSON_CONFIG.debugEnabled && cache.level !== 0) {
        console.error(`RichJson mergeIntoTarget failed!`);
    }
    return target;
}

function _mergeIntoTarget(cache, target, others) {
    let other = undefined;

    cache.level++;
    for (let i = 0; i < others.length; i++) {
        other = others[i];
        if (other === undefined) {
            continue;
        }
        __mergeIntoTarget(cache, target, other);
    }
    cache.level--;

    return target;
}

export function __mergeIntoTarget(cache, target, other, force = false) {
    let names = Object.keys(other);
    let name = undefined;
    let member = undefined;

    for (let j = 0; j < names.length; j++) {
        name = names[j];
        member = other[name];

        if (typeof member === "function") {
            if (!force && !Object.hasOwn(target, name))  {
                target[name] = member.bind(target);
            }
        } else {
            if (member !== undefined && isJsonObject(member) && isJsonObject(target[name])) {
                if (!Object.hasOwn(cache.stack, cache.resolveAddress(member))) {
                    cache.stack[cache.resolveAddress(member)] = member;
                    _mergeIntoTarget(cache, target[name], [member]);
                }
            } else if (!force && !Object.hasOwn(target, name)) {
                target[name] = member;
            }
        }
    }

    return target;
}

/**
 * Merges two or more objects together into a new object, without rebinding functions.
 * @param {object} objects Going to be merged
 * @returns {object} The merged object
 */
export function mergeObjectsWithoutRebind(objects) {
    let rv = {};
    let cache = new RichJsonCache();
    cache.stack = [];
    for (let i = 0; i < objects.length; i++) {
        _mergeIntoWithoutRebind(cache, rv, [objects[i]]);
    }
    if (__RICH_JSON_CONFIG.debugEnabled && cache.level !== 0) {
        console.error(`RichJson mergeIntoTarget failed!`);
    }
    return rv;
}

/**
 * Merges objects into target object, without rebinding functions.
 * @param {object} target
 * @param {object} others Going to be merged
 * @returns {object} The merged target
 */
export function mergeIntoWithoutRebind(target, ...others) {
    let cache = new RichJsonCache();
    cache.stack = [];
    target = _mergeIntoWithoutRebind(cache, target, others);
    if (__RICH_JSON_CONFIG.debugEnabled && cache.level !== 0) {
        console.error(`RichJson mergeIntoWithoutRebind failed!`);
    }
    return target;
}

function _mergeIntoWithoutRebind(cache, target, others) {
    let other = undefined;

    cache.level++;
    for (let i = 0; i < others.length; i++) {
        other = others[i];
        if (other === undefined) continue;
        __mergeIntoWithoutRebind(cache, target, other);
    }
    cache.level--;

    return target;
}

function __mergeIntoWithoutRebind(cache, target, other, force = false) {
    let names = Object.keys(other);
    let name = undefined;
    let member = undefined;

    for (let j = 0; j < names.length; j++) {
        name = names[j];
        member = other[name];
        if (typeof member === "function") {
            if (!force && !Object.hasOwn(target, name)) target[name] = member;
        } else {
            if (member !== undefined &&
                isJsonObject(member) && isJsonObject(target[name]) &&
                !cache.stack.includes(member)) {
                cache.stack.push(member);
                _mergeIntoWithoutRebind(cache, target[name], [member]);
            } else {
                if (!force && !Object.hasOwn(target, name)) target[name] = member;
            }
        }
    }

    return target;
}

/**
 * Clones an object or array. Circular dependencies are correctly resolved and functions are rebound.
 * @param {object} object
 * @returns {object} The clone
 */
export function cloneObject(object) {
    let cache = new RichJsonCache();
    cache.str = isJsonObject(object);
    cache.rv = (isJsonObject(object) ? {} : []);
    cache.next = undefined;
    cache.next = cache.rv;
    cache.stack[cache.resolveAddress(object)] = cache.rv;
    object = _cloneObject(cache, object);
    if (__RICH_JSON_CONFIG.debugEnabled && cache.level !== 0) {
        console.error(`RichJson cloneObject failed!`);
    }
    return object;
}

function _cloneObject(cache, object) {
    if (object == null || (!isJsonObject(object) && !Array.isArray(object))) {
        return object;
    }

    let target = cache.next;
    let name = undefined;
    let member = undefined;
    let newObj = undefined;

    cache.level++;
    if (isJsonObject(object)) {
        let names = Object.keys(object);
        for (let i = 0; i < names.length; i++) {
            name = names[i];
            member = object[name];
            if (typeof member === "function") {
                target[name] = member.bind(target);
            } else if (isJsonObject(member) || Array.isArray(member)) {
                if (getFieldByKey(cache.stack, cache.resolveAddress(member)) === undefined) {
                    newObj = (isJsonObject(member)) ? {} : [];
                    cache.str = isJsonObject(member);
                    cache.stack[cache.resolveAddress(member)] = newObj;
                    cache.next = newObj;
                    target[name] = _cloneObject(cache, member);
                } else {
                    target[name] = cache.stack[cache.resolveAddress(member)];
                }
            } else {
                target[name] = member;
            }
        }
    } else {
        for (let i = 0; i < object.length; i++) {
            member = object[i];
            if (isJsonObject(member) || Array.isArray(member)) {
                if (getFieldByKey(cache.stack, cache.resolveAddress(member)) === undefined) {
                    newObj = (isJsonObject(member)) ? {} : [];
                    cache.str = isJsonObject(member);
                    cache.stack[cache.resolveAddress(member)] = newObj;
                    cache.next = newObj;
                    target.push(_cloneObject(cache, member));
                } else {
                    target.push(cache.stack[cache.resolveAddress(member)]);
                }
            } else {
                target.push(member);
            }
        }
    }
    cache.level--;

    return target;
}

/**
 * Gets a field of an object by key.
 * @param {object} object
 * @param {string} key
 * @returns {object} object[key] - If object has field
 * @returns {undefined} undefined - If object has not
 */
export function getFieldByKey(object, key) {
    return Object.hasOwn(object, key) ? object[key] : undefined;
}

/**
 * Gets keys sorted of given object.
 * @param {object} object
 * @returns {String[]} The keys
 */
export function getKeysSorted(object) {
    let rv = Object.keys(object);
    rv.sort((a, b) => a.localeCompare(b, undefined, {sensitivity: "base"}));
    return rv;
}

/**
 * Checks if given object is a JSON object.
 * @param {object} object
 * @returns {boolean}
 */
export function isJsonObject(object) {
    return typeof object === 'object' && !Array.isArray(object) && (object.constructor === Object || Object.hasOwn(__RICH_JSON_CLASS_MAPPING, object.constructor.name));
}

/**
 * Concat redundancy strings.
 * @param {...String} strings
 * @returns {String}
 */
export function concatStrings(...strings) {
    let rv = "";
    strings.forEach((str) => rv += str);
    return rv;
}

/**
 * Concat redundancy arrays.
 * @param {...Array} arrays
 * @returns {Array}
 */
export function concatArrays(...arrays) {
    let rv = [];
    arrays.forEach((array) => rv.push(array));
    return rv;
}

/**
 * Checks if the given string matches the wildcard.
 * @param string
 * @param wildcard
 * @returns {boolean}
 */
export function matchesWildcard(string, wildcard) {
    if (typeof string !== "string") return false;
    const escaped = wildcard.replace(/[.+?^${}()|\[\]\\]/g, "\\$&");
    const regexStr = "^" + escaped.replace(/\*/g, ".*") + "$";
    wildcard = new RegExp(regexStr);
    return wildcard.test(string);
}

File: src/java/main/RichJsonHelper.java
import java.util.*;
import java.util.regex.Pattern;

public class RichJsonHelper {

    // ==========================================
    // STRING & ARRAY UTILS
    // ==========================================

    public static String concatStrings(String... strings) {
        return String.join("", strings);
    }

    @SafeVarargs
    public static List<Object> concatArrays(List<Object>... arrays) {
        List<Object> result = new ArrayList<>();
        for (List<Object> array : arrays) {
            if (array != null) {
                result.addAll(array);
            }
        }
        return result;
    }

    public static boolean matchesWildcard(String string, String wildcard) {
        if (string == null || wildcard == null) return false;
        // Escaped alle Regex-Zeichen außer dem Sternchen (*)
        String escaped = wildcard.replaceAll("[.+?^${}()|\\[\\]\\\\]", "\\\\$0");
        String regexStr = "^" + escaped.replace("*", ".*") + "$";
        return string.matches(regexStr);
    }

    // ==========================================
    // OBJECT UTILS
    // ==========================================

    public static boolean isJsonObject(Object object) {
        // Da wir Jackson nutzen, sind JSON-Objekte immer Maps.
        return object instanceof Map;
    }

    public static Object getFieldByKey(Object object, String key) {
        if (object instanceof Map) {
            return ((Map<?, ?>) object).get(key);
        }
        return null;
    }

    @SuppressWarnings("unchecked")
    public static List<String> getKeysSorted(Object object) {
        if (!(object instanceof Map)) return new ArrayList<>();
        List<String> keys = new ArrayList<>(((Map<String, Object>) object).keySet());
        // Sortierung Case-Insensitive (Basis-Sensitivität im JS)
        keys.sort(String.CASE_INSENSITIVE_ORDER);
        return keys;
    }

    // ==========================================
    // MERGE LOGIC
    // ==========================================

    /**
     * Entspricht mergeObjects und mergeObjectsWithoutRebind
     */
    @SafeVarargs
    public static Map<String, Object> mergeObjects(Map<String, Object>... objects) {
        return mergeIntoTarget(new HashMap<>(), objects);
    }

    /**
     * Entspricht mergeIntoTarget und mergeIntoWithoutRebind
     */
    @SafeVarargs
    public static Map<String, Object> mergeIntoTarget(Map<String, Object> target, Map<String, Object>... others) {
        for (Map<String, Object> other : others) {
            if (other == null) continue;

            RichJsonCache cache = new RichJsonCache();
            __mergeIntoTarget(cache, target, other, false);

            if (cache.level != 0) {
                System.err.println("RichJson mergeIntoTarget failed!");
            }
        }
        return target;
    }

    @SuppressWarnings("unchecked")
    public static Map<String, Object> __mergeIntoTarget(RichJsonCache cache, Map<String, Object> target, Map<String, Object> other, boolean force) {
        String address = cache.resolveAddress(other);
        cache.stack.put(address, other);
        cache.level++;

        for (Map.Entry<String, Object> entry : other.entrySet()) {
            String name = entry.getKey();
            Object member = entry.getValue();

            if (member != null && isJsonObject(member) && isJsonObject(target.get(name))) {
                if (!cache.stack.containsKey(cache.resolveAddress(member))) {
                    __mergeIntoTarget(cache, (Map<String, Object>) target.get(name), (Map<String, Object>) member, force);
                }
            } else if (force || !target.containsKey(name)) {
                target.put(name, member);
            }
        }

        cache.level--;
        return target;
    }

    // ==========================================
    // CLONE LOGIC
    // ==========================================

    /**
     * Klont Maps, Listen und Primitives tief (Deep Clone). Löst Zirkelreferenzen auf.
     */
    public static Object cloneObject(Object object) {
        RichJsonCache cache = new RichJsonCache();
        Object cloned = _cloneObject(cache, object);
        if (cache.level != 0) {
            System.err.println("RichJson cloneObject failed!");
        }
        return cloned;
    }

    @SuppressWarnings("unchecked")
    private static Object _cloneObject(RichJsonCache cache, Object object) {
        // Null oder Primitives (String, Integer, etc.) werden direkt zurückgegeben
        if (object == null || (!isJsonObject(object) && !(object instanceof List))) {
            return object;
        }

        String address = cache.resolveAddress(object);

        // Zirkelreferenz entdeckt! Gebe die bereits geklonte Referenz zurück
        if (cache.stack.containsKey(address)) {
            return cache.stack.get(address);
        }

        cache.level++;
        Object newObj;

        if (isJsonObject(object)) {
            Map<String, Object> sourceMap = (Map<String, Object>) object;
            Map<String, Object> targetMap = new HashMap<>();
            newObj = targetMap;

            // Sofort im Stack speichern für Kind-Knoten, die auf uns zurückzeigen könnten
            cache.stack.put(address, targetMap);

            for (Map.Entry<String, Object> entry : sourceMap.entrySet()) {
                targetMap.put(entry.getKey(), _cloneObject(cache, entry.getValue()));
            }
        } else {
            List<Object> sourceList = (List<Object>) object;
            List<Object> targetList = new ArrayList<>();
            newObj = targetList;

            cache.stack.put(address, targetList);

            for (Object item : sourceList) {
                targetList.add(_cloneObject(cache, item));
            }
        }

        cache.level--;
        return newObj;
    }
}